// habemus-editor-services/config/index.js
// native dependencies
const url = require('url');

// third-party dependencies
const Bluebird = require('bluebird');
const pathToRegExp = require('path-to-regexp');

// h-deps
const HAccountDialog = require('h-account-client/browser/ui/dialog');

// injected modules
const habemusEditorUrls = require('habemus-editor-urls');

/**
 * Demo mode
 * TODO: deprecate demo mode on launch
 */
const DEMO_MODE_ENABLED = process.env.DEMO_MODE === 'enabled';
function _loadDemoData(habemus, options) {
  console.log('loading account from demo configs');
  var parsedURL = url.parse(window.location.toString(), true);
  var demoData = parsedURL.query.demo;
  
  // use the demo data as auth token
  habemus.services.hAccountDialog.hAccountClient._saveAuthToken(demoData);

  delete parsedURL.search;
  delete parsedURL.query.demo;
  
  // remove the demo data from the url
  window.history.replaceState(
    'habemus-editor',
    'Habemus',
    url.format(parsedURL)
  );
}

// http://stackoverflow.com/questions/19999388/check-if-user-is-using-ie-with-jquery
function _isMSInternetExplorer() {
  var ua = window.navigator.userAgent;
  var msie = ua.indexOf("MSIE ");

  if (msie !== -1 || !!navigator.userAgent.match(/Trident.*rv\:11\./)) {
    return true;
  } else {
    return false;
  }
}
const IS_IE = _isMSInternetExplorer();

function loadConfig(habemus, options) {
  console.log('load configs for `browser-cloud` environment');
  
  if (IS_IE) {
    
    var msg = habemus.services.criticalLanguage.t('workspace-setup.error.browser-not-supported');
    
    habemus.services.dialogs.alert(msg);
    
    return Bluebird.reject(new Error('browser not supported'));
  }

  var locationString = window.location.toString();

  // get project code from the url
  var projectCode = habemusEditorUrls.parse.uiWorkspace(locationString).projectCode;
  if (!projectCode) {
    var err = new Error('could not parse projectCode');

    // error is equivalent to `NotFound`
    err.name = 'NotFound';

    throw err;
  }
  
  /**
   * DEMO MODE
   */
  if (DEMO_MODE_ENABLED && url.parse(locationString, true).query.demo) {
    _loadDemoData(habemus, options);
  }
  
  return habemus.services.hAccountDialog.ensureAccount({
    ensureEmailVerified: true,
  })
  .then(function (account) {
    
    return {
      apiVersion: '1.0.0',
      projectId: projectCode,
      projectName: projectCode,

      // cloud specific options
      cloud: {
        apiVersion: '1.0.0',
        hWorkspaceURI: process.env.H_WORKSPACE_URI,
        uiDashboardURI: process.env.UI_DASHBOARD_URI,
        projectCode: projectCode,
      },

      language: require('./language')(account, habemus, options),
      
      /**
       * browser-cloud only projectMenu options
       */
      projectMenuOptions: require('./project-menu-options')(account, habemus, options),
    };
  });
}

module.exports = loadConfig;
