// habemus-editor-services/config/project-menu-options.js
// injected modules
const habemusEditorURLs = require('habemus-editor-urls');

// TODO: study how to merge with language configs
const AVAILABLE_LANGUAGES = [
  {
    label: 'English (US)',
    key: 'en-US',
  },
  {
    label: 'Português (BR)',
    key: 'pt-BR',
  },
  {
    label: '中文 (简体)',
    key: 'zn-CH',
  }
];

module.exports = function (account, habemus, options) {
  return [
    {
      label: 'publish new version',
      callback: function (data) {
        data.menuElement.close();
        var context = data.context;
        
        console.log(context);
        
        var config = habemus.services.config;
        var hDev   = habemus.services.hDev;
        var authToken = habemus.services.hAccountDialog.getAuthToken();
        
        /**
         * Show the loading notification
         * and let it be manually removed
         */
        habemus.services.notification.loading.show({
          text: 'Publising a new version of "' + config.cloud.projectCode + '"',
          duration: Math.Infinity,
        });
    
        return hDev.createProjectVersion(
          authToken,
          config.cloud.projectCode, {
            byProjectCode: true,
          }
        )
        .then(function (workspace) {
          /**
           * Hide loading notification
           */
          habemus.services.notification.loading.hide();
          habemus.services.notification.success.show({
            text: 'Published version "' + workspace.projectVersionCode + '"',
            duration: 7000,
          });
        })
        .catch(function (err) {
    
          habemus.services.notification.loading.hide();
          habemus.services.notification.error.show({
            text: 'An unexpected error occurred: ' + err.name,
            duration: 7000, 
          });
    
          console.warn('create version error', err);
        });
      }
    },
    
    {
      label: 'view in production',
      type: 'url',
      url: function (data) {
        
        var config = habemus.services.config;
        
        return habemusEditorURLs.format.websiteHabemusDomain(
          config.cloud.projectCode
        );
      },
      target: '_blank',
      callback: function (data) {
        data.menuElement.close();
      },
    },
    
    {
      label: 'languages',
      type: 'submenu',
      options: AVAILABLE_LANGUAGES.map((language) => {
        return {
          label: language.label,
          callback: function (data) {
            data.menuElement.close();
            
            var hAccountDialog = habemus.services.hAccountDialog;
            
            /**
             * Save the language configuration
             */
            hAccountDialog.getCurrentAccount()
              .then(function (account) {
                return hAccountDialog.hAccountClient.updateApplicationConfig(
                  hAccountDialog.getAuthToken(),
                  account.username,
                  'workspace',
                  {
                    language: language.key,
                  }
                );
              });
        
            /**
             * Reset application language
             */
            habemus.services.language.resetLanguage(language.key)
              .catch(function (err) {
                // the error only happens whenever the user cancels the prompt.
                // The prompt is used to prevent the user from refreshing the 
                // page while there are unsaved changes, but the
                // language setting might be considered done anyway.
              });
          }
        }
      }),
    },
    {
      label: 'logout',
      callback: function (data) {
        data.menuElement.close();
        
        habemus.services.dialogs.confirm('All unsaved changes will be lost. Are you sure you want to logout?', {
          confirm: 'logout',
        })
        .then(function () {
          return habemus.services.hAccountDialog.logOut();
        })
        .then(function () {
          // TODO: replace with env var.
          window.location.assign('https://habem.us');
        })
        .catch(function (err) {
          if (err.canceled) {
            // user canceled
          } else {
            console.warn('logout error');
            console.warn(err);
          }
        });
      },
    }
  ]
};
