// hb-service-config/index.js
// native dependencies
const url = require('url');

// third-party dependencies
const Bluebird = require('bluebird');

const pkg = require('../../../package.json');

function loadConfig() {
  console.log('load configs for browser environment');

  return new Bluebird((resolve, reject) => {

    var locationString = window.location.toString();
    var parsedURL      = url.parse(locationString, true);

    if (!parsedURL.query || !parsedURL.query.projectId) {
      throw new Error('`projectId` must be provided as query parameter');
    }

    var projectId = parsedURL.query.projectId;
    var authToken = localStorage.getItem('habemus_auth_token');

    if (!authToken) {
      throw new Error('`habemus_auth_token` must be set in localStorage');
    }

    resolve({
      apiVersion: pkg.version,
      hDevServerURI: 'http://localhost:5001',
      // hDevServerURI: 'http://104.196.140.133',
      authToken: authToken,
      projectId: parsedURL.query.projectId,
      projectName: parsedURL.query.projectId,

      // cloud specific options
      cloud: {
        apiVersion: pkg.version,
        // hDevServerURI: 'http://localhost:5001',
        hDevServerURI: 'http://104.196.140.133',
        authToken: authToken,
      }
    });
  });
}

module.exports = loadConfig;
