// hb-service-config/index.js
// native dependencies
const url = require('url');

// third-party dependencies
const Bluebird = require('bluebird');
const pathToRegExp = require('path-to-regexp');

// h-deps
const HAccountDialog = require('h-account-client/browser/ui/dialog');

const pkg = require('../../../package.json');

const DEFAULT_WORKSPACE_ROUTE = '/workspace/:workspaceCode';

const H_ACCOUNT_URI = process.env.H_ACCOUNT_URI;
const H_WORKSPACE_URI = process.env.H_WORKSPACE_URI;
const WORKSPACE_ROUTE = process.env.WORKSPACE_ROUTE || DEFAULT_WORKSPACE_ROUTE;

const workspaceRegExp = pathToRegExp(WORKSPACE_ROUTE);

function parseWorkspaceCodeFromLocation() {
  var locationString = window.location.toString();
  var parsedURL      = url.parse(locationString, true);

  var code;

  // first attempt to get the url from the path
  var workspaceRouteMatch = parsedURL.path.match(workspaceRegExp);

  if (workspaceRouteMatch) {
    code = workspaceRouteMatch[1];
  }

  // if no code is found in the path,
  // get it from the query string
  code = code || parsedURL.query.code;

  return code;
}

function loadConfig() {
  console.log('load configs for browser environment');

  if (!H_ACCOUNT_URI) {
    throw new Error('H_ACCOUNT_URI env var MUST be set');
  }
  if (!H_WORKSPACE_URI) {
    throw new Error('H_WORKSPACE_URI env var MUST be set');
  }

  var locationString = window.location.toString();

  var workspaceCode = parseWorkspaceCodeFromLocation();

  if (!workspaceCode) {
    throw new Error('could not parse workspaceCode');
  }

  // instantiate hAuth
  var accountDialog = new HAccountDialog({
    serverURI: H_ACCOUNT_URI,
    containerElement: document.body,
  });

  return accountDialog.ensureUser()
    .then(function (user) {

      var authToken = accountDialog.getAuthToken();

      if (!authToken) {
        throw new Error('could not retrieve authToken');
      }

      return {
        apiVersion: pkg.version,
        projectId: workspaceCode,
        projectName: workspaceCode,

        // cloud specific options
        cloud: {
          apiVersion: pkg.version,
          hWorkspaceURI: H_WORKSPACE_URI,
          authToken: authToken,
          workspaceCode: workspaceCode,
        }
      };
    })
    .catch(function (err) {
      console.warn(err);
    });
}

module.exports = loadConfig;
