// service-hfs/index.js
// third-party dependencies
const Bluebird   = require('bluebird');


module.exports = function (options) {

  const hfs = {
    readDirectory: function (p) {
      console.log('readDirectory', p);
      return Bluebird.resolve([
        {
          basename: 'file',
          isDirectory: function () {
            return false;
          }
        },
        {
          basename: 'file-2',
          isDirectory: function () {
            return false;
          }
        },
        {
          basename: 'folder',
          isDirectory: function () {
            return true;
          }
        },
      ]);
    },

    remove: function (p) {
      console.log('remove', p);
      return Bluebird.resolve();
    },

    writeFile: function (p, contents) {
      console.log('writeFile', p, contents);
      return Bluebird.resolve();
    },

    move: function (src, dest) {
      console.log('move', src, dest);
      return Bluebird.resolve();
    },

    readFile: function (p, options) {
      console.log('readFile', p, options);
      return Bluebird.resolve(p + ' contents');
    },

    publish: function (eventName, eventData) {
      // console.log('event ', eventName, eventData);
      return Bluebird.resolve();
    },
    

    createDirectory: function (p) {
      console.log('createDirectory', p)
      return Bluebird.resolve();
    },


    createFile: function (p) {
      console.log('createFile', p);
      return Bluebird.resolve();
    },
  };

  return Bluebird.resolve(hfs);
}
