// habemus-editor-services/config.js
// native dependencies
const path = require('path');

// third-party dependencies
const Bluebird = require('bluebird');

const pkg = require('../../../../package.json');

const CONSTANTS = require('../../../../src/constants');

function loadConfig(habemus, options) {
  // retrieve the FS_ROOT_PATH
  // it defaults to process.cwd()
  const FS_ROOT_PATH = process.cwd();

  console.log('load configs for `electron` environment');

  return {
    apiVersion: pkg.version,
    
    // required by core
    projectId: FS_ROOT_PATH,
    projectName: path.basename(FS_ROOT_PATH),

    // electron specific options
    electron: {
      fsRootPath: FS_ROOT_PATH,
    },

    language: {
      baseURL: process.env.LANGUAGES_BASE_URL || '/resources/languages',

      fallback: 'en-US',

      languages: [
        {
          label: 'English (US)',
          key: 'en-US',
          data: require('../../../../src/resources/languages/en-US.json'),
        },
        {
          label: 'Português (BR)',
          key: 'pt-BR',
        },
        {
          label: '中文 (简体)',
          key: 'zn-CH',
        }
      ],
    }
  };
}

module.exports = loadConfig;
